### ISSUE TITLE
File times don't return fractional info (nsec, usec, subsec)

### ISSUE THREAD

**Issue by MSP-Greg** [MSP-Greg/ruby_on_windows#1]
@kubo

No, thank you.  I just added the patch to a folder and clicked a cmd file, then checked the logs.  You wrote code.

On Win7, all three (.atime, .ctime, & .mtime) return .nsec, .usec, and .subsec.

I hate to ask, but in trying to add some tests back in for windows builds, it seems that [File.utime](https://msp-greg.github.io/ruby_trunk/Core/File.html#utime-class_method) (click '[ view_source ]') doesn't set the fractional part of atime and mtime.  Is that possible?

Origin - [oneclick/rubyinstaller2 #52](https://github.com/oneclick/rubyinstaller2/issues/52)

**Comment by kubo** [comment]
Could you try [this patch](https://gist.githubusercontent.com/kubo/0a351c5d6fbbd3d3f7cce44a0fbed9d9/raw/d242f5d6bc269c500a2a4d17f09b3924fffbf4a3/02-file-utime.dif)?
It must be applied after the patch posted at https://github.com/oneclick/rubyinstaller2/issues/52#issuecomment-312282894.
The rightmost two digits of nanosecond are truncated to zeros because the timestamp resolution of NTFS is 100-nanosecond.

**Comment by MSP-Greg** [comment]
@kubo 

Wonderful.  Built fine, tested fine, my test code for it shows utime functioning now.

On Win7, I can't tell what the OS is doing to atime.  Also, I'm not seeing any resolution beyond 100 microseconds (0.1 ms, or 4 digits) for any of the fractional methods.

But, utime now changes both atime and mtime, with fractional seconds.

Tomorrow (I'm -0500) I'll go thru the Ruby tests and enable the tests for the utime issue.  Once I get that done, maybe you (or I) can PR it to ruby.  I'm interested in seeing if it passes the mswin build, which now has full testing done on it.

Thanks again.

BTW, I thought I'd just make a repo for issues/discussions about Ruby on Windows.  The discussion was moving in a direction that had nothing to do with RubyInstaller2...

**Comment by kubo** [comment]
On Windows 10 (version 1703), the resolution is 100 nanoseconds as follows:
```ruby
require 'tempfile'

class Time
  def to_s_with_nsec
    format('%d-%02d-%02d %02d:%02d:%02d.%09d',
           self.year, self.mon, self.day,
           self.hour, self.min, self.sec,
           self.nsec)
  end
end

utime = Time.new(2012, 3, 4, 5, 6, 7_123_456_789.to_r / 1_000_000_000)
tmp = Tempfile.new('foo')
File.utime(utime, utime, tmp.path)
mtime1 = File.mtime(tmp.path)
atime1 = File.atime(tmp.path)
mtime2 = tmp.stat.mtime
atime2 = tmp.stat.atime
tmp.close!

puts 'utime:  ' + utime.to_s_with_nsec
puts 'mtime1: ' + mtime1.to_s_with_nsec
puts 'atime1: ' + atime1.to_s_with_nsec
puts 'mtime2: ' + mtime2.to_s_with_nsec
puts 'atime2: ' + atime2.to_s_with_nsec
```
This prints the following:
```
utime:  2012-03-04 05:06:07.123456789
mtime1: 2012-03-04 05:06:07.123456700
atime1: 2012-03-04 05:06:07.123456700
mtime2: 2012-03-04 05:06:07.123456700
atime2: 2012-03-04 05:06:07.123456700
```

**Comment by MSP-Greg** [comment]
Yes, you're correct.  Files times have 100 ns resolution when set via code (.utime).

I was using Time.now for my .utime value, it seems Win7 has 100 µs resolution for that.  I'll start looking thru those tests soon.  Thanks again for your work.

**Comment by MSP-Greg** [comment]
@kubo 

After a diversion into creating a script for x64-mswin64_140 build/test, I just opened [issue 13726](https://bugs.ruby-lang.org/issues/13726) on ruby with your patches.  I've build and tested them on x64-mingw32 and x64-mswin64_140.

Thanks again.

**Comment by kubo** [comment]
@MSP-Greg 

Thank you for posing the issue!

**Off-topic:**

> I was using Time.now for my .utime value, it seems Win7 has 100 µs resolution for that.

The resolution seems 15.6 msec (1/64 second) by default. It depends on timer resolution in the system.
```ruby
timer_resolution = nil

if timer_resolution
  require 'win32api'
  timeBeginPeriod = Win32API.new('winmm.dll', 'timeBeginPeriod', 'l', 'l')
  if timeBeginPeriod.call(timer_resolution) != 0
    puts "timeBeginPeriod failed."
  end
end

res = []
prev = Time.now
loop_cnt = 0
while res.length < 10
  loop_cnt += 1
  t = Time.now
  if t != prev
    res.push(t - prev)
    prev = t
  end
end
puts "loop #{loop_cnt} times"
puts res.join("\n")
puts "min: #{res.min}"
puts "max: #{res.max}"
puts "avg: #{res.inject(&:+) / res.length}"
```
This script printed:
```
loop 188999 times
0.015568
0.015626
0.015626
0.015627
0.015625
0.015626
0.015627
0.015626
0.015627
0.015626
min: 0.015568
max: 0.015627
avg: 0.015620400000000001
```
The resolution of Time.now seems 15.6 ms. (Stop multimedia applications in advance, which may change timer resolution in the system.)

When the first line of the script was changed to `timer_resolution = 1` (1 msec), it printed:
```
loop 11806 times
0.000999
0.001001
0.001001
0.001001
0.001001
0.001001
0.001
0.001002
0.001001
0.001
min: 0.000999
max: 0.001002
avg: 0.0010007000000000002
```
The resolution was changed to 1.0 ms.

**Comment by MSP-Greg** [comment]
@kubo 

I reviewed my testing script, and, you're correct, Time resolution is not 100 µs.

What I did see is that when windows natively writes file times, it seems to write them with 100 µs resolution.  But, Ruby can write with higher resolution using utime.

**Comment by kubo** [comment]
@MSP-Greg 
I posted precise Time.now patch to [issue 13732 in Japanese](https://bugs.ruby-lang.org/issues/13732).
Time.now resolution is 1 µs with the patch when the Windows version is Windows 8 (Windows Server 2012) or later.

**Comment by MSP-Greg** [comment]
@kubo 

Thanks again for all your help.  As I think I've mentioned, I'm not a c type...

I've got probably a really dumb question.  I recall that you've mentioned that the set of patches wasn't ABI compatible.  Your code looked like it redefined some functions/methods/structures because the previous ones weren't being used anymore.  Hence, the stupid question - if some of the old/previous functions/methods/structures were left defined, could it also be 'back ABI compatible'?

You've done enough work, so I'm not asking you to do so.

For quite a while, I've only been building/testing trunk, but recently I started doing the same for 2.3.4, ruby_2_3 stable, 2.4.1, and ruby_2_4 stable.  There are some failures / errors I've seen before on trunk.

So, there have been updates that the core team did for issues found in MinGW trunk that I'm hoping can  be backported.  At the time, I wasn't testing, so I didn't think of mentioning it...

Thought I'd double check re the time patches you've done.  Thanks again, and sorry for the english...
